cloudsync/event.py

```diff
@@ -292,26 +297,52 @@ def _fill_event_path(self, event: Event):
                 event.otype = info.otype
                 event.hash = info.hash
     
-    def _filter_event(self, event: Event) -> bool:
+    def _filter_event(self, event: Event, from_walk: bool = False) -> bool:
         # event filtering based on root path
+        # return True = ignore event (filter it out), False = process event
         if not self._root_path:
             return False
-        # path not a subpath of root
-        if event.exists and event.path and not self.provider.is_subpath(self._root_path, event.path):
-            lookup_oid = event.prior_oid or event.oid
-            state = self.state.lookup_oid(self.side, lookup_oid)
-            if state:
-                # found in state, so this was moved out of our root -- delete
-                log.debug("change event to delete - moved from relevance to irrelevance: %s", state[self.side].path)
-                event.exists = False
+
+        # detect root changes
+        if self.provider.root_oid == event.oid:
+            if not event.exists:
+                raise CloudRootMissingError(f"root was deleted for provider: {self.provider.name}")
+            if not self.provider.paths_match(self.provider.root_path, event.path):
+                raise CloudRootMissingError(f"root was renamed for provider: {self.provider.name}")
+        if self.provider.root_oid == event.prior_oid:
+            raise CloudRootMissingError(f"root was renamed for provider: {self.provider.name}")
+
+        # filter for create/modify
+        if event.exists and event.path:
+            subpath = self.provider.is_subpath_of_root(event.path)
+            state = self.state.lookup_oid(self.side, (event.prior_oid or event.oid))
+
+            if state and not subpath:
+                # item was moved out of root_path -- treated as delete
+                if event.otype == DIRECTORY and not from_walk:
+                    log.debug("directory renamed out of root - walking to pick up events for children")
+                    self._do_walk_oid(event.oid)
+                log.debug("moved from relevance to irrelevance: %s", state[self.side].path)
```

Comment:
No way it's okay to trigger a walk every single time we move a file out of a folder